### PR TITLE
Ashforb/dim app refactor

### PR DESCRIPTION
**Pull request by forbesashli**
### Summary
Changed the 9-segments to be written via Drive State, via groups of three. Tests updated to reflect this.
### Changelist
\boards\DIM\ Src\main.c
\boards\DIM\Src\App\states\App_DriveState.c
\boards\DIM\Src\App\App_SevenSegDisplay.c
\boards\DIM\Src\App\App_SevenSegDisplays.c
\boards\DIM\Src\Io\Io_SevenSegDisplays.c

\boards\DIM\Inc\App\App_SevenSegDisplay.h
\boards\DIM\Inc\App\App_SevenSegDisplays.h
\boards\DIM\Test\Src\Test_StateMachine.cpp
\boards\DIM\Test\Src\Test_SevenSegDisplays.cpp

### Testing Done
Re-written tests to reflect the change to the three group writing. Tested correct number outputs, re-writing one segment does not touch others and overflow/underflow error handling. This mainly tests the App code
Tested these features on the DIM board, get expected outputs with inputs.
### Resolved Issues
Do not believe it resolves any perscribes issues, only re-works the 7 segments to write to 9 as opposed to 3
### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: boards/DIM/Inc/App/App_SevenSegDisplay.h [hunk 1]
```diff
@@ -18,11 +18,12 @@ struct SevenSegHexDigit
  * @return A pointer to the created 7-segment display, whose ownership is given
  *         to the caller
  */
-struct SevenSegDisplay *App_SevenSegDisplay_Create(void (*set_hex_digit)(struct SevenSegHexDigit));
+struct SevenSegDisplay *App_SevenSegDisplay_Create(void (*set_hex_digit)(struct SevenSegHexDigit, int DIGIT));
```
Comment (gtaharaedmonds) [inline boards/DIM/Inc/App/App_SevenSegDisplay.h:21]: variables should be `snake_case`

File: boards/DIM/Inc/App/App_SevenSegDisplay.h [hunk 2]
```diff
@@ -18,11 +18,12 @@ struct SevenSegHexDigit
...
 
 /**
  * Deallocate the memory used by the given 7-segment display
  * @param seven_seg_display The 7-segment display to deallocate
+ * @param digit
```
Comment (gtaharaedmonds) [inline boards/DIM/Inc/App/App_SevenSegDisplay.h:26]: looks like this is in the wrong spot

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 3]
```diff
@@ -37,74 +51,119 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
-    // If any of the input digits is invalid, we don't write anything to the
-    // 7-segment displays at all.
-    for (size_t i = 0; i < num_hex_digits; i++)
+    for (size_t i = 0; i < len_hex_digits; i++)
     {
         if (hex_digits[i] >= NUM_HEX_DIGITS)
         {
             return EXIT_CODE_INVALID_ARGS;
         }
     }
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:65]: Does this function do anything?

Comment (gtaharaedmonds) [reply boards/DIM/Src/App/App_SevenSegDisplays.c:65]: Also, can use `NUM_SEVEN_SEG_DISPLAYS` for length

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 4]
```diff
@@ -37,74 +51,119 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
 
-    for (size_t i = 0; i < NUM_SEVEN_SEG_DISPLAYS; i++)
+    return EXIT_CODE_OK;
+}
+
+void App_Set_Digits(const struct SevenSegDisplays *seven_seg_displays, uint8_t index)
+{
+    for (int i = index; i < NUM_IN_GROUP + index; i++)
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:72]: Functions name should be `(App/Io)_(file name)_(function name)`
So `App_SevenSegDisplays_SetDigits` 

The name `index` is pretty vague, consider renaming to something like `group_index` so it's clear this refers to index of group, not index of digit or something else. Also could make a new enum of the 3 groups and pass that instead.

You should add some error checking too, if someone passed a 4 for ex. the program would crash.

Also, it doesn't look like function will be used in any other files? If so it should be removed from the header file and declared statically in this .c file.

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 5]
```diff
@@ -37,74 +51,119 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
+
+ExitCode App_SevenSegDisplays_SetGroupL(const struct SevenSegDisplays *const seven_seg_displays, uint32_t value)
+{
+    if (value > pow(10, NUM_IN_GROUP) - 1)
+    {
+        return EXIT_CODE_INVALID_ARGS;
+    }
+    digits[0] = 0;
+    digits[1] = 0;
+    digits[2] = 0;
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:92]: Can replace these indices with the digit enum values. In general I think magic numbers like this are usually a bad idea

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 6]
```diff
@@ -37,74 +51,119 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
+
+    // Turn the base-10 value into individual digits. Have to write backwards with how the
+    // displays are initialized and how they are passed to the IO function.
+    int shift = 3;
+    for (int digits_index = shift; digits_index + shift > NUM_IN_GROUP; digits_index--)
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:97]: This shift variable seems a big magic too, maybe something like this says what you're doing more clearly
`for (int digit = LEFT_R_SEVEN_SEG_DISPLAY; digit >= LEFT_L_SEVEN_SEG_DISPLAY; digits_index--)`

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 7]
```diff
@@ -37,74 +51,119 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
+        value /= 10;
+
+        if (value == 0)
         {
-            // We turn off the 7-segment displays with unspecified values. For
-            // example, if the callers wants to write 0xF, we would turn off the
-            // right and middle 7-segment displays.
-            hex_digit.enabled = false;
+            App_Set_Digits(seven_seg_displays, 0);
+            break;
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:105]: Can the call to `App_Set_Digits` go after the loop? Are there cases where value != 0 after the last run

File: boards/DIM/Src/App/states/App_DriveState.c [hunk 8]
```diff
@@ -3,95 +3,6 @@
...
-{
-    // No DIM Faults yet
-    return false;
-}
-
-static bool App_DriveState_HasFSMFault()
-{
-    // TODO: Promote this message to fault
-    return App_CanRx_FSM_Warnings_MissingHeartbeat_Get();
-}
```
Comment (gtaharaedmonds) [inline boards/DIM/Src/App/states/App_DriveState.c:27]: Ooh looks like your git got a little silly

Comment (gtaharaedmonds) [reply boards/DIM/Src/App/states/App_DriveState.c:27]: You should try to revert all the files which you didn't intentionally edit
You can checkout single files from another branch: https://stackoverflow.com/questions/2364147/how-to-get-just-one-file-from-another-branch
If you do that and commit and push again it shouldn't diff anything here

File: can-bus/dbcs/CanMsgs.dbc [hunk 9]
```diff
@@ -0,0 +1,921 @@
...
+
+
+BO_ 502 DIM_Vitals: 1 DIM
+SG_ Heartbeat : 0|1@1+ (1,0) [0|1] "" DEBUG,BMS
+SG_ State : 1|1@1+ (1,0) [0|0] "" DEBUG,BMS
+
+BO_ 503 DIM_Switches: 1 DIM
+SG_ StartSwitch : 0|1@1+ (1,0) [0|1] "" DEBUG
+SG_ AuxSwitch : 1|1@1+ (1,0) [0|1] "" DEBUG
+
```
Comment (gtaharaedmonds) [inline can-bus/dbcs/CanMsgs.dbc:16]: we don't track dbc on git anymore, this file can be deleted

**Review by gtaharaedmonds — CHANGES_REQUESTED** [review]
Looks great, just a few nitpicky things 🦧

File: can-bus/dbcs/CanMsgs.dbc (hunk 9 above)
Comment (forbesashli) [reply can-bus/dbcs/CanMsgs.dbc:16]: done

File: boards/DIM/Src/App/App_SevenSegDisplays.c (hunk 6 above)
Comment (forbesashli) [reply boards/DIM/Src/App/App_SevenSegDisplays.c:97]: done

File: boards/DIM/Src/App/App_SevenSegDisplays.c (hunk 7 above)
Comment (forbesashli) [reply boards/DIM/Src/App/App_SevenSegDisplays.c:105]: done

File: boards/DIM/Src/App/App_SevenSegDisplays.c (hunk 4 above)
Comment (forbesashli) [reply boards/DIM/Src/App/App_SevenSegDisplays.c:72]: Also, it doesn't look like function will be used in any other files? If so it should be removed from the header file and declared statically in this .c file.


Naming convention fixed, 

Not really sure if I understand what you mean by passing a 4, I don't think it'll crash, just edit the incorrect places in the digit array. The error checking is done before this function is called for incorrect numbers, as was the fashion of the old code so I stuck with it believing it was convention (?)

File: boards/DIM/Inc/App/App_SevenSegDisplay.h (hunk 2 above)
Comment (forbesashli) [reply boards/DIM/Inc/App/App_SevenSegDisplay.h:26]: done

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 10]
```diff
@@ -37,74 +52,103 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
 
-    // If any of the input digits is invalid, we don't write anything to the
-    // 7-segment displays at all.
-    for (size_t i = 0; i < num_hex_digits; i++)
+    // Turn the base-10 value into individual digits. Have to write backwards with how the
+    // displays are initialized and how they are passed to the IO function.
+    for (int digits_index = LEFT_R_SEVEN_SEG_DISPLAY; digits_index + LEFT_R_SEVEN_SEG_DISPLAY > NUM_IN_GROUP; digits_index--)
     {
-        if (hex_digits[i] >= NUM_HEX_DIGITS)
+        digits[digits_index - 1] = (uint8_t)(value % 10);
```
Comment (1mackenziekyle) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:83]: This was brought up by Gus, but just leaving as a reminder: digits is indexed by [digits_index - 1] here, but is indexed by [digits_index] in SetGroupM and SetGroupR

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 11]
```diff
@@ -37,74 +52,103 @@ void App_SevenSegDisplays_Destroy(struct SevenSegDisplays *const seven_seg_displ
...
 
-        if (i < num_hex_digits)
-        {
-            hex_digit.enabled = true;
-            hex_digit.value   = hex_digits[i];
-        }
-        else
+    // Turn the base-10 value into individual digits. Have to write backwards with how the
+    // displays are initialized and how they are passed to the IO function
+    for (int digits_index = MIDDLE_R_SEVEN_SEG_DISPLAY; digits_index + MIDDLE_R_SEVEN_SEG_DISPLAY > NUM_IN_GROUP; digits_index--)
```
Comment (1mackenziekyle) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:109]: Regarding code style, might be more human-readable to directly index the `digits` array using the ints 0-8, and simply write which index of the array corresponds to which location on the display. 

For example: SetGroupL might look like
for (i = 0; i < 3; i++) // do stuff to digits[i]

Whereas SetGroupM might be 
for (i = 3; i < 6; i++) // do stuff to digits[i]

Comment (gtaharaedmonds) [reply boards/DIM/Src/App/App_SevenSegDisplays.c:109]: I'm usually against unnamed constants but I can see the enums here is unclear
Gonna leave this one up to Ashli

File: boards/DIM/Src/Io/Io_SevenSegDisplays.c [hunk 12]
```diff
@@ -74,15 +65,14 @@ void Io_SevenSegDisplays_WriteCommands(void)
...
     {
-        // TODO: Update commands for 9 displays
-        const uint8_t display_data = formula_e_commands[TEMP_NUM_DISPLAYS - display - 1]; // invert order
+        const uint16_t display_data = commands[9 - display - 1]; // invert order
 
         for (int i = 0; i < SHIFT_REGISTER_SIZE; i++)
         {
             const bool bit_state = IS_BIT_SET(display_data, i);
-
+            // TODO correct pinouts
```
Comment (will-chaba) [inline boards/DIM/Src/Io/Io_SevenSegDisplays.c:75]: is this still outstanding?

If this is to be pushed to another PR, use // TODO: syntax so it gets labelled as a todo item in Clion

File: boards/FSM/Src/main.c [hunk 13]
```diff
@@ -24,42 +24,6 @@
...
-#include "App_FsmWorld.h"
-#include "states/App_DriveState.h"
-
-// Sensors
-#include "App_AcceleratorPedals.h"
-#include "App_Brake.h"
-#include "App_Coolant.h"
-#include "App_Steering.h"
-#include "App_Wheels.h"
-#include "configs/App_HeartbeatMonitorConfig.h"
```
Comment (will-chaba) [inline boards/FSM/Src/main.c:62]: Uhhhh I don't think this was intentional, please undo all changes to FSM/Src/main.c

File: boards/shared/Inc/App/App_SharedMacros.h [hunk 14]
```diff
@@ -40,7 +40,21 @@
...
+
+// Convert inches to kilometers
+#define INCH_TO_KM (0.0000254f)
+
 #define RPM_TO_RADS(rpm) ((rpm) * (float)M_PI / 30.0f)
 
 #define PI (3.1415926f)
 #define RAD_TO_DEG(r) (r * 180.0f / PI)
+
+#define MOTOR_RPM_TO_KMH(kmh) ((kmh) * (float)WHEEL_DIAMETER*PI*INCH_TO_KM*MIN_TO_HOUR/GEAR_RATIO) //take rpm of whell to kph
```
Comment (will-chaba) [inline boards/shared/Inc/App/App_SharedMacros.h:60]: nit: missing newline

**Review by will-chaba — CHANGES_REQUESTED** [review]
Accidentally deleted changes to FSM main.c, please fix

File: boards/DIM/Src/App/App_SevenSegDisplays.c [hunk 15]
```diff
@@ -24,87 +53,166 @@ struct SevenSegDisplays *App_SevenSegDisplays_Create(
...
+        exponent = 2;
+    }
+    else if (value >= 10){
+        exponent = 1;
+    }
+
+    uint16_t display_value = (uint16_t)(value * power_lookup_table.values[exponent]);;
+
+    // Turn the base-10 value into individual digits. Have to write backwards with how the
+    // displays are initialized and how they are passed to the IO function.
```
Comment (kironsen) [inline boards/DIM/Src/App/App_SevenSegDisplays.c:120]: this reused code could be its own static function

**Review by will-chaba — APPROVED** [review]
Thanks for those fixes, looks good to me
